go/store/nbs/table_index.go

```diff
@@ -0,0 +1,506 @@
+	// |Close|d independently.
+	Clone() (tableIndex, error)
+}
+
+func ReadTableFooter(rd io.ReadSeeker) (chunkCount uint32, totalUncompressedData uint64, err error) {
+	footerSize := int64(magicNumberSize + uint64Size + uint32Size)
+	_, err = rd.Seek(-footerSize, io.SeekEnd)
+
+	if err != nil {
+		return 0, 0, err
+	}
+
+	footer, err := iohelp.ReadNBytes(rd, int(footerSize))
+
+	if err != nil {
+		return 0, 0, err
+	}
+
+	if string(footer[uint32Size+uint64Size:]) != magicNumber {
+		return 0, 0, ErrInvalidTableFile
+	}
+
+	chunkCount = binary.BigEndian.Uint32(footer)
+	totalUncompressedData = binary.BigEndian.Uint64(footer[uint32Size:])
+
+	return
+}
+
+// parses a valid nbs tableIndex from a byte stream. |buff| must end with an NBS index
+// and footer, though it may contain an unspecified number of bytes before that data.
+// |tableIndex| doesn't keep alive any references to |buff|.
+// Does not allocate new memory except for offsets, computes on buff in place.
+func parseTableIndex(buff []byte) (onHeapTableIndex, error) {
+	chunkCount, totalUncompressedData, err := ReadTableFooter(bytes.NewReader(buff))
+	if err != nil {
+		return onHeapTableIndex{}, err
+	}
+
+	iS := indexSize(chunkCount)
+	buff = buff[:len(buff)-footerSize]
+	// Trim away any extra bytes
+	buff = buff[uint64(len(buff))-iS:]
+
+	return NewOnHeapTableIndex(buff, chunkCount, totalUncompressedData)
+}
+
+// ReadTableIndex loads an index into memory from an io.ReadSeeker
+// Caution: Allocates new memory for entire index
+func ReadTableIndex(rd io.ReadSeeker) (onHeapTableIndex, error) {
+	chunkCount, totalUncompressedData, err := ReadTableFooter(rd)
+	if err != nil {
+		return onHeapTableIndex{}, err
+	}
+	iS := int64(indexSize(chunkCount))
+	_, err = rd.Seek(-(iS + footerSize), io.SeekEnd)
+	if err != nil {
+		return onHeapTableIndex{}, ErrInvalidTableFile
+	}
+	buff := make([]byte, iS)
+	_, err = io.ReadFull(rd, buff)
+	if err != nil {
+		return onHeapTableIndex{}, err
+	}
+
+	return NewOnHeapTableIndex(buff, chunkCount, totalUncompressedData)
+}
+
+type onHeapTableIndex struct {
+	tableFileSize uint64
+	// Tuple bytes
+	tupleB []byte
+	// Offset bytes
+	offsetB []byte
+	// Suffix bytes
+	suffixB               []byte
+	chunkCount            uint32
+	totalUncompressedData uint64
+}
+
+var _ tableIndex = &onHeapTableIndex{}
+
+// NewOnHeapTableIndex creates a table index given a buffer of just the table index (no footer)
+func NewOnHeapTableIndex(b []byte, chunkCount uint32, totalUncompressedData uint64) (onHeapTableIndex, error) {
+	tuples := b[:prefixTupleSize*chunkCount]
+	lengths := b[prefixTupleSize*chunkCount : prefixTupleSize*chunkCount+lengthSize*chunkCount]
+	suffixes := b[prefixTupleSize*chunkCount+lengthSize*chunkCount:]
+
+	lR := bytes.NewReader(lengths)
+	offsets := make([]byte, chunkCount*offsetSize)
+	_, err := io.ReadFull(NewOffsetsReader(lR), offsets)
+	if err != nil {
+		return onHeapTableIndex{}, err
+	}
+
```

Comment:
Probably worth adding a TODO here about how `len(lengths)` in the backing array is over retained and we could technically squeeze offsets into that space + another allocation of size `len(lengths)`.